### PR TITLE
[33648] Add value macros for referencing label and values of project and work packages

### PR DESCRIPTION
**Pull request by oliverguenther**
**TODOS**
  - [x] Add component and macro for referencing labels with attribute help texts of projects and work packages
  - [x] Add component and macro for referencing values of projects and work packages
  - [x] Add component for work package quicklinks `##ID` and `##ID` with varying level of details (tbd) 
  - [x] Tests for parsing/outputting the markdown macro
  - [x] Feature spec
  - [x] Extend [docs](https://github.com/opf/openproject/pull/8578/commits/48ba86d7b5a1104b768234fcca0d93d65e4c98e0?short_path=b6697bd#diff-b6697bd71d03e7162a99819b9be7eb88)

https://community.openproject.com/wp/33648
https://community.openproject.com/wp/33525
https://community.openproject.com/wp/33962

### CODE REVIEW

File: docs/user-guide/wysiwyg/README.md [hunk 1]
```diff
@@ -129,10 +129,79 @@ As with the textile formatting syntax, you can link to other resources within Op
...
+| Remaining Time      | `workPackageValue:1234:remainingTime`                        |
+| Responsible user    | `workPackageValue:1234:responsible`                          |
+| Spent time          | `workPackageValue:1234:spentTime`                            |
+| Start date          | `workPackageValue:1234:startDate`                            |
+| Status              | `workPackageValue:1234:status`                               |
+| Subject / Title     | `workPackageValue:1234:subject`                              |
+| Work package type   | `workPackageValue:1234:type`                                 |
+| Date of last update | `workPackageValue:1234:updatedAt`                            |
+| Version             | `workPackageValue:1234:version`                              |
+| *Custom Fields*     | `workPackageValue:1234:"Name of the work package custom field"` |
```
Comment (ulferts) [inline docs/user-guide/wysiwyg/README.md:190]: There are actually additional attributes that stem from modules, e.g. budget, overallCosts. However, while overallCosts work, budget fails silently.

**Review by ulferts — CHANGES_REQUESTED** [review]
The following irritates me as a user:

* Only resources of the current project can be referenced. I believe this to be an unnecessary restriction. Especially as the same restriction does not apply to the `##` and `###` marcos.

* When entering a `###` macro for a non existing work package, an empty <macro> element is added, which is different from the behaviour of inserting a non existing attribute.

* When entering a macro for a non existing project, e.g. `projectValue:bogus:name` an empty <macro> element is added, which is different from the behaviour of inserting a non existing attribute.

* I cannot activate the edit field by clicking on a macro. As it is not apparent, until having activated the edit field by clicking somewhere else, where a marco provided the value, this looks as if it where a bug. 

* It might be useful to better visualize when text was added by a macro. Otherwise, I might want to fix/alter something I only realize to be editable elsewhere after having already clicked to edit the text. 

* The work package autocompleter does not work when having typed `##` or `###` 

* Expanding some custom field values via the macro did not work, e.g. a custom field "cf_text" fails without a user visible notification.

* Probably negligible: When specifying "costsByType" as the attribute of the macro, the list elements are preceded by '-' 

The following irritates me as a developer:

* Every macro fetches its resource individually (although caching prevents fetching the same resource multiple times). Especially when the marco will be used in some sort of report/list, which I imagine to be the main use case of the feature, this will mean having references to a lot of different resources on a page which all need to be fetched.

**Comment by oliverguenther** [comment]
Thanks for looking into this.

> Only resources of the current project can be referenced. I believe this to be an unnecessary restriction. Especially as the same restriction does not apply to the ## and ### marcos.

**Fixed** :heavy_check_mark:  You should be able reference work packages of other projects by using their ID. I however made a mistake in detecting only numbers, which results in never matching the number test. https://github.com/opf/openproject/commit/4012b46836 fixed this.

Using the subject, only the current project will be searched. This reduces the possibility to find other resources. For city of cologne use case, this would happen all the time since the phases are named identically.


> When entering a ### macro for a non existing work package, an empty element is added, which is different from the behaviour of inserting a non existing attribute.
> When entering a macro for a non existing project, e.g. projectValue:bogus:name an empty element is added, which is different from the behaviour of inserting a non existing attribute.

**Fixed** :heavy_check_mark:  Now shows an error when not found similar to the other macros

> I cannot activate the edit field by clicking on a macro. As it is not apparent, until having activated the edit field by clicking somewhere else, where a marco provided the value, this looks as if it where a bug.

This check is by design when clicking on larger macro blocks (embedded table?). https://github.com/opf/openproject/blob/ac12a00177b86d87cda5a69dcec76f43a5ff1af1/frontend/src/app/modules/fields/edit/field/editable-attribute-field.component.ts#L164 We could try to differentiate between kind of macros. However if we make the macro more visible, it would maybe become clear what is not editable?

> It might be useful to better visualize when text was added by a macro. Otherwise, I might want to fix/alter something I only realize to be editable elsewhere after having already clicked to edit the text.

I'm open for ideas but we'll need to discuss this with city of cologne as the values should likely be shown as if entered manually which conflicts with this.

> The work package autocompleter does not work when having typed ## or ###

:x: **won't fix**  I tried to make multiple mentions available with a marker for `##` and `###` but the plugin only accepts single char markers: https://ckeditor.com/docs/ckeditor5/latest/framework/guides/support/error-codes.html#error-mentionconfig-incorrect-marker

> Expanding some custom field values via the macro did not work, e.g. a custom field "cf_text" fails without a user visible notification.

:question: **Can't reproduce** Could you double check? I just tested this again with a user CF and got errors if it was not available, and it was correctly output.

> Probably negligible: When specifying "costsByType" as the attribute of the macro, the list elements are preceded by '-'

**Fixed** :heavy_check_mark:  Found two more cases where this happens, this appears to be a bug in the display field not overriding the placeholder. It doesn't show when using the `editable-attribute-field` since that clears the innerHTML property. Will look into it


> Every macro fetches its resource individually (although caching prevents fetching the same resource multiple times).

:x: **won't fix** As I need to rely on a query filter to look for the value, I'm not sure how this would be resolvable. Work packages referenced by their IDs can be aggregated, but not by their subject.

/Edit Looks like the subject filter uses `contains` operator which [looks whether the string matches all provided array values](https://github.com/opf/openproject/blob/e5749d9817c60cd04e05b16e5285f7cd79e7ec3d/app/models/queries/operators/concerns/contains_all_values.rb#L41)

**Review by ulferts — APPROVED** [review]
Good to merge upon Travis' consent.

Could not reproduce the cf reference problems myself anymore so I assume that it no longer exists/never existed/had to do with my setup...

**Pull request merged**